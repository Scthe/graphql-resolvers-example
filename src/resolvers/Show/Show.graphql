extend type Query {
  # TODO add query by imdb
  show(id: ID!): Show!
  shows(name: String): ShowsList!
}

type Show {
  id: ID!
  name: String!
  type: String! # TODO enum?
  language: String!
  genres: [String!]!
  status: String! # TODO enum
  runtime: Int!
  premiered: String! # TODO date
  #network: String
  #country: String
  summary: String!
  # TODO imdb
  seasons: SeasonsList! # Paginated collection with total
}

type ShowsList {
  node: [Show!]!
  meta: ListMeta!
}
