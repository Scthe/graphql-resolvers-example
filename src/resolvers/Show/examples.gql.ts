export const GET_SINGLE_GQL = `{
  show(id: 118){
    id
    name
    seasons{
      meta {
        totalCount
      }
      node{
        id
        idx
        summary
        show {
          name
        }
      }
    }
  }
}`;

export const GET_LIST_GQL = `{
  shows(name: "House"){
    meta{
      totalCount
    }
    node{
  		id
  	  name
    }
  }
}`;