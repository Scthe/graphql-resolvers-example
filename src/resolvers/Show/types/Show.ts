import GqlContext from "GqlContext";
import { Show } from "typingsGql";
import { BaseResolverType, copyFromRestResponse } from "utils/graphql";

import { RootType as SeasonsListType } from "../../Season/types/SeasonsList";
import { RootType as ShowCharactersListType } from "../../ShowCharacter/types/ShowCharactersList";

export type RootType = ID;

type ResolverType = BaseResolverType<
  RootType,
  Show,
  {
    id: ID;
    seasons: SeasonsListType;
    cast: ShowCharactersListType;
  }
>;

const getItem = (id: RootType, context: GqlContext) => {
  return context.dataSources.showsAPI.getOne(id);
};

const imdbId = async (
  root: RootType,
  _args: any,
  context: GqlContext
): Promise<string> => {
  const item = await getItem(root, context);
  return item.externals.imdb;
};

const resolver: ResolverType = {
  id: (root: RootType) => root,
  name: copyFromRestResponse(getItem, "name"),
  type: copyFromRestResponse(getItem, "type"),
  language: copyFromRestResponse(getItem, "language"),
  genres: copyFromRestResponse(getItem, "genres"),
  status: copyFromRestResponse(getItem, "status"),
  runtime: copyFromRestResponse(getItem, "runtime"),
  premiered: copyFromRestResponse(getItem, "premiered"),
  // network: copyFromRestResponse(getItem, "network"),
  // country: copyFromRestResponse(getItem, "country"),
  summary: copyFromRestResponse(getItem, "summary"),
  imdbId,
  seasons: (root: RootType) => ({ showId: root }),
  cast: (root: RootType) => ({ showId: root }),
};

export default resolver;
