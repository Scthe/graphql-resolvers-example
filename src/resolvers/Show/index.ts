import { IResolvers } from "apollo-server";

import ShowQuery from "./queries/Show";
import ShowByImdbQuery from "./queries/ShowByImdb";
import ShowsQuery from "./queries/Shows";
import Show from "./types/Show";
import ShowsList from "./types/ShowsList";

export default {
  Query: {
    show: ShowQuery,
    shows: ShowsQuery,
    showByIMDB: ShowByImdbQuery,
  },
  Show,
  ShowsList,
} as IResolvers<any, any>; // eslint-disable-line @typescript-eslint/no-explicit-any
