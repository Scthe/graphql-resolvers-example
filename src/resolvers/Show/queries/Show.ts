import { QueryShowArgs } from "typingsGql";
import { RootType as ShopType } from "../types/Show";

export default (_root: any, args: QueryShowArgs): ShopType => parseInt(args.id, 10) as any;
