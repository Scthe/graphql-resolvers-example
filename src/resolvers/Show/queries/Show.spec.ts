import supertest from "supertest";
import nock, { Scope } from "nock";

import app from "../../../index";
import * as gqlExamples from "./Show.example";
import { Show as GqlSchemaShow } from "typingsGql";

import ShowMock from "dataSources/tvmaze/mocks/show.mock.json";
import SeasonMock from "dataSources/tvmaze/mocks/season.mock.json";
import EpisodeMock from "dataSources/tvmaze/mocks/episode.mock.json";

describe("GraphQL: show()", () => {
  let scope: Scope | null = null;

  beforeEach(() => {
    // each jest test loads modules anew
    nock.disableNetConnect();
    nock.enableNetConnect("127.0.0.1");
  });

  afterEach(() => {
    scope && scope.done(); // verify all planned requests were actually done
    nock.cleanAll();
  });

  afterAll(() => {
    nock.restore(); // https://github.com/nock/nock/issues/1817
  });

  const mockSuccessShowResponse = (item = ShowMock) => {
    // NOTE: this allows for ONE request
    scope = nock("https://api.tvmaze.com")
      .get("/shows/118")
      .reply(200, item);
  };

  const mockSuccessSeasonsResponse = () => {
    const seasonMock = (id: string, number: number) => ({
      ...SeasonMock,
      id,
      number,
    });
    const episodeMock = (id: number, name: string) => ({
      ...EpisodeMock,
      id,
      name,
    });

    scope!
      .get("/shows/118/seasons")
      .reply(200, [
        seasonMock("123", 1),
        seasonMock("999", 2),
      ])
      .get("/seasons/123/episodes")
      .reply(200, [
        episodeMock(1231, "season-123-episode-1"),
        episodeMock(1232, "season-123-episode-2"),
      ])
      .get("/seasons/999/episodes")
      .reply(200, [
        episodeMock(9991, "season-999-episode-1"),
      ]);
  };

  test("should only fetch what is needed", () => {
    // In this test we check if we overfetch. We query only for id (which we pass in args),
    // so no network request is expected (nock would throw).
    return supertest(app)
      .post("/graphql")
      .send({
        query: gqlExamples.GQL_SINGLE_SHOW_ONLY_ID,
      })
      .then(resp => {
        const item = resp.body.data;
        expect(item.show.id).toBe("118");
        expect(item).toMatchSnapshot(); // just for a good measure
      })
  });

  test("should fetch basic item properties", () => {
    // In this test we get all properties of a show (that are not a paginated list).
    // We should be able to deduplicate request and do it only once.
    mockSuccessShowResponse();

    return supertest(app)
      .post("/graphql")
      .send({
        query: gqlExamples.GQL_SHOW_FIELDS,
      })
      .then(resp => {
        const item = resp.body.data;
        expect(item.show.id).toBe("" + ShowMock.id);
        expect(item).toMatchSnapshot(); // just for a good measure
      })
      .finally();
  });

  test("should fetch show with all children depdendencies", () => {
    // This is quite extreme test. We will fetch show with all seasons and episodes.
    // In production app this test would be VERY brittle, so be careful.
    // What we want is to invoke as much resolvers as we can (show/season/episode).
    // TBH I'm doing this to showcase that all connections in graph work correctly.
    mockSuccessShowResponse();
    mockSuccessSeasonsResponse();

    return supertest(app)
      .post("/graphql")
      .send({
        query: gqlExamples.GQL_SHOW_WITH_CHILDREN,
      })
      .then(resp => {
        const item = resp.body.data.show as GqlSchemaShow;
        expect(item.id).toBe("" + ShowMock.id);
        expect(item.seasons.meta.totalCount).toBe(2);
        expect(item.seasons.node).toHaveLength(2);
        // check season 1
        expect(item.seasons.node[0].id).toBe("123");
        expect(item.seasons.node[0].episodes.node).toHaveLength(2);
        expect(item.seasons.node[0].episodes.meta.totalCount).toBe(2);
        expect(item.seasons.node[0].episodes.node[0].id).toBe("1231");
        expect(item.seasons.node[0].episodes.node[1].id).toBe("1232");
        // check season 2
        expect(item.seasons.node[1].id).toBe("999");
        expect(item.seasons.node[1].episodes.node).toHaveLength(1);
        expect(item.seasons.node[1].episodes.meta.totalCount).toBe(1);
        expect(item.seasons.node[1].episodes.node[0].id).toBe("9991");

        // just for a good measure
        expect(item).toMatchSnapshot();
      })
      .finally();
  });
});
