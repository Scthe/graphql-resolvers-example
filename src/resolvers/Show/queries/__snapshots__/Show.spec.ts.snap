// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GraphQL: show() should fetch basic item properties 1`] = `
Object {
  "show": Object {
    "genres": Array [
      "Drama",
      "Mystery",
      "Medical",
    ],
    "id": "118",
    "language": "English",
    "name": "House",
    "premiered": "2004-11-16",
    "runtime": 60,
    "status": "Ended",
    "summary": "<p>Sink your teeth into meaty drama and intrigue with <b>House</b>, FOX's take on mystery, where the villain is a medical malady and the hero is an irreverent, controversial doctor who trusts no one, least of all his patients.</p>",
    "type": "Scripted",
  },
}
`;

exports[`GraphQL: show() should fetch show with all children depdendencies 1`] = `
Object {
  "id": "118",
  "name": "House",
  "seasons": Object {
    "meta": Object {
      "totalCount": 2,
    },
    "node": Array [
      Object {
        "episodes": Object {
          "meta": Object {
            "totalCount": 2,
          },
          "node": Array [
            Object {
              "id": "1231",
              "name": "season-123-episode-1",
              "season": Object {
                "id": "123",
                "idx": 1,
              },
              "show": Object {
                "id": "118",
                "name": "House",
              },
            },
            Object {
              "id": "1232",
              "name": "season-123-episode-2",
              "season": Object {
                "id": "123",
                "idx": 1,
              },
              "show": Object {
                "id": "118",
                "name": "House",
              },
            },
          ],
        },
        "id": "123",
        "idx": 1,
        "show": Object {
          "id": "118",
          "name": "House",
        },
      },
      Object {
        "episodes": Object {
          "meta": Object {
            "totalCount": 1,
          },
          "node": Array [
            Object {
              "id": "9991",
              "name": "season-999-episode-1",
              "season": Object {
                "id": "999",
                "idx": 2,
              },
              "show": Object {
                "id": "118",
                "name": "House",
              },
            },
          ],
        },
        "id": "999",
        "idx": 2,
        "show": Object {
          "id": "118",
          "name": "House",
        },
      },
    ],
  },
}
`;

exports[`GraphQL: show() should only fetch what is needed 1`] = `
Object {
  "show": Object {
    "id": "118",
  },
}
`;
